File: src/main/java/com/github/aarcangeli/ideaclangformat/ClangFormatResponse.java
package com.github.aarcangeli.ideaclangformat;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.dataformat.xml.XmlMapper;
import com.fasterxml.jackson.dataformat.xml.annotation.JacksonXmlElementWrapper;
import com.fasterxml.jackson.dataformat.xml.annotation.JacksonXmlProperty;
import com.fasterxml.jackson.dataformat.xml.annotation.JacksonXmlRootElement;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;

@JacksonXmlRootElement(localName = "replacements")
public class ClangFormatResponse {

  public static ClangFormatResponse unmarshal(@NotNull String stdout) {
    try {
      XmlMapper xmlMapper = new XmlMapper();
      return xmlMapper.readValue(stdout, ClangFormatResponse.class);
    }
    catch (JsonProcessingException e) {
      throw new RuntimeException(e);
    }
  }

  @JacksonXmlProperty(localName = "space", isAttribute = true)
  public String space;

  @JacksonXmlProperty(localName = "incomplete_format", isAttribute = true)
  public boolean incompleteFormat;

  @JacksonXmlProperty(localName = "replacement")
  @JacksonXmlElementWrapper(useWrapping = false)
  public final List<ClangFormatReplacement> replacements = new ArrayList<>();
}

File: src/main/kotlin/com/github/aarcangeli/ideaclangformat/services/ClangFormatResponse.kt
package com.github.aarcangeli.ideaclangformat.services

import java.io.StringReader
import javax.xml.bind.JAXBContext
import javax.xml.bind.JAXBException
import javax.xml.bind.annotation.XmlAttribute
import javax.xml.bind.annotation.XmlElement
import javax.xml.bind.annotation.XmlRootElement
import javax.xml.bind.annotation.XmlValue

fun parseClangFormatResponse(stdout: String): ClangFormatResponse {
  val ctx: JAXBContext?
  try {
    ctx = JAXBContext.newInstance(ClangFormatResponse::class.java)
  }
  catch (e: JAXBException) {
    throw RuntimeException("Failed to load JAXB context", e)
  }

  val result = ctx.createUnmarshaller().unmarshal(StringReader(stdout))
  return result as ClangFormatResponse
}

@XmlRootElement(name = "replacements")
class ClangFormatResponse {
  @set:XmlElement(name = "replacement")
  var replacements: List<ClangFormatReplacement> = ArrayList()
}

class ClangFormatReplacement {
  @set:XmlAttribute
  var offset: Int = 0

  @set:XmlAttribute
  var length: Int = 0

  @set:XmlValue
  var value: String = ""
}
